### DISCUSSION TITLE
Source git vs tarball + spec

### DISCUSSION THREAD

**Discussion by dcermak** [openSUSE/openSUSE-git#3] [General]
I have started reading https://github.com/aplanas/openSUSE-git/blob/main/notes-about-git.md and it looks like you're proposing what I know as the "source git" approach, i.e. the packager clones the upstream git repo, includes the spec and adds additional patches as commits on top. This approach is used by Debian as well and was also discussed in Fedora where the [Source Git SIG](https://fedoraproject.org/wiki/SIGs/Source-git) is currently working on refining this further.

While this approach has some upsides (rebasing patches is easy and feels very natural), I would highly encourage us to not tread down this path for the following reasons:
1. it is very easy to add downstream patches and don't contribute them upstream. Especially it is rather hard to add some tooling that checks how many patches you are applying and raise a red flag. Historically there were a few offenders in openSUSE that had **hundreds** of downstream patches and this would make their life easier, which is wrong imho: patching **must** be hard. You should be forced to put everything upstream as much as possible, because maintaining patches downstream is not sustainable.
2. force-pushing is required for version bumps. This is imho the major nogo of the source git approach: your own downstream changes and the upstream version bump get mixed up during a version bump PR and make the diff extremely noisy. Additionally, you loose the history and need to resort to hacks to keep previous revisions (that were force pushed away) around. My major gripe with this is however, that it becomes very hard to view the evolution of your spec and of downstream patches over time. That is imho the important part of keeping the packaging recipes (spec + patches) in version control: to be able to view changes and with source git, you sacrifice this.

Thus I would like to propose to abandon this approach and resort to the "classic" dist-git like approach, i.e. spec + patches are in git and the upstream sources are stored as an archive (created by whatever means). This works perfectly with a completely linear history, it allows you to track how many patches exist and how they evolved over time and it is also compatible with the current packaging mental model. Most importantly, you can forbid force pushing completely, which turns your distribution branches into append-only logs ensuring that no one can inject something nasty and hide their traces by force pushing it away.

**Comment by lnussel** [comment]
>     1. it is very easy to add downstream patches and don't contribute them upstream. Especially it is rather hard to add some tooling that checks how many patches you are applying and raise a red flag. Historically there were a few offenders in openSUSE that had **hundreds** of downstream patches and this would make their life easier, which is wrong imho: patching **must** be hard.

On the contrary. It must be really easy to be able to quickly adjust software to openSUSE. Having the full package in git then also makes it quick and easy to submit fixes directly to upstream. The annoying method we have now just lead to creative scripting to maintain tons of patches nevertheless while those who want to work in cooperation with upstream git have to suffer.

> You should be forced to put everything upstream as much as possible, because maintaining patches downstream is not sustainable.

That should be a matter of policy and reviews rather than forcing inferior workflows on packagers. Certainly there needs to be document that describes how to do it right. That's already missing with todays tooling.

**Reply by dcermak** [reply]
> On the contrary. It must be really easy to be able to quickly adjust software to openSUSE. Having the full package in git then also makes it quick and easy to submit fixes directly to upstream. The annoying method we have now just lead to creative scripting to maintain tons of patches nevertheless while those who want to work in cooperation with upstream git have to suffer.

I highly doubt that doing a `git format patch` is so severely limiting your ability to push updates. And it's not that I disagree with you that this additional step is annoying and feels pointless, I just don't agree that it outweighs the downsides.

Also I'd like to add my own experience here: in vagrant I have a bunch of downstream patches (usually rubygem version bumps + non-upstreamable patches & test skips) and I maintain these in git. However, upstream will most certainly never accept my specfile into their repo, so I have to add another commit for that. Then most upstreamable patches have to be submitted from separate branches as separate pull requests. In a nutshell, I have to cherry-pick a bunch of commits anyway and for me personally the benefit of having everything in the same git repo would be not overwhelmingly helpful. It would remove one step but would cost me a rather clear overview of how my spec file & patches evolved over time, which I would hate to loose.

**Comment by aplanas** [comment]
> I have started reading https://github.com/aplanas/openSUSE-git/blob/main/notes-about-git.md and it looks like you're proposing what I know as the "source git" approach, i.e. the packager clones the upstream git repo, includes the spec and adds additional patches as commits on top.

Note that there are two proposals, and both must coexist in the same timeline. One is the traditional layout (tarball + spec), and the other one is the one that you are referring to. Tarballs will stay for a very long time, if not for ever, as could be that the best decision for certain packages.

> https://fedoraproject.org/wiki/SIGs/Source-git

I will add the link into the document.

>     1. it is very easy to add downstream patches and don't contribute them upstream.

Uau no, just the opposite. In the example the remote upstream is explicitly added in the user branch. The developer will maintain the patches and remove then transparently once they are merged upstream. As a mater of fact, it is using the same tool and process to maintain the upstream patches and the downstream ones. It is merging both worlds, making trivial the contribution.

> Especially it is rather hard to add some tooling that checks how many patches you are applying and raise a red flag.

There is no difference. The git cli make easy to count the number of commits living in one branch wrt the upstream one. If this metric is required, can be done in one line of git (not tested, but can be something like `git log --oneline Tumbleweed ^upstream | wc -l`)
 
>     2. force-pushing is required for version bumps. 

Again, no difference. Is exactly the same as "a new tarball requires a version bumps". Is equally true and equally tractable. 

> This is imho the major nogo of the source git approach: your own downstream changes and the upstream version bump get mixed up during a version bump PR and make the diff extremely noisy. 

Is the same that updating a tarball and rebasing the patches with quilt. The difference here is that the `git log` is making explicit the real changes, meanwhile in the tarball version is what the user say in the changes file.

> Additionally, you loose the history and need to resort to hacks to keep previous revisions (that were force pushed away) around.

I do not understand this. As showed later in the doc I presented a model where no history get lost.

> My major gripe with this is however, that it becomes very hard to view the evolution of your spec and of downstream patches over time.

That is a good point, but in the proposal those are stored in the ref/build/XX branches. They are never collected, and available directly on those branches and / or in the revlog. Again, nothing get lost. But is a good point that needs to be clarified.

> Thus I would like to propose to abandon this approach and resort to the "classic" dist-git like approach, i.e. spec + patches are in git and the upstream sources are stored as an archive (created by whatever means).

As commented, this model should be present. But IMHO the git model is better in any metric. IIUC your points are addressed in the current draft.

**Reply by dcermak** [reply]
> Uau no, just the opposite. In the example the remote upstream is explicitly added in the user branch. The developer will maintain the patches and remove then transparently once they are merged upstream. As a mater of fact, it is using the same tool and process to maintain the upstream patches and the downstream ones. It is merging both worlds, making trivial the contribution.

Please take the lazy/busy contributor into account. If their patches are directly in git and updating to the next upstream release is just a `git rebase upstream/next`, then the pain of maintaining a downstream patch is very small but it **must be**. Maintaining downstream patches must be as painful as possible, so that you try to get rid of them ASAP. This might sound harsh, but imho that's the only way how to force packagers to not diverge too hard from upstream.

> There is no difference. The git cli make easy to count the number of commits living in one branch wrt the upstream one. If this metric is required, can be done in one line of git (not tested, but can be something like `git log --oneline Tumbleweed ^upstream | wc -l`)

That only works if you know the upstream git repositories' tag/branch naming scheme (and there's only about 50 different naming conventions). That could become much more complicated than the current `grep -i '^patch:' *spec|wc`.

> Again, no difference. Is exactly the same as "a new tarball requires a version bumps". Is equally true and equally tractable.

I don't understand how. If you have downstream patches on top of `upstream/1.0`, you rebase them on top of `upstream/1.1` and you push this to your `tumbleweed` branch, then you **must** force push.

**Reply by aplanas** [reply]
> Maintaining downstream patches must be as painful as possible

I get the point, but it is against what this effort is about. I advocate for a model like this one to resolve the issues that I had when managing patches.

But if that was the goal, I think the succeeded long ago.

> > Again, no difference. Is exactly the same as "a new tarball requires a version bumps". Is equally true and equally tractable.
> 
> I don't understand how. If you have downstream patches on top of `upstream/1.0`, you rebase them on top of `upstream/1.1` and you push this to your `tumbleweed` branch, then you **must** force push.

Indeed. Pbly I misunderstood your point then.

Reading again your comment I think that the main idea that you want to make is that the force push (in case that this is the recommended policy) will make the history get lost. That is correct, but the solution for this issue is to have those `ref/build/XX` branches created by OBS. In that way we can:

* The "Tumbleweed" branch will represent the upstream changes, and stacked on top the downstream changes (patches and spec file).  A new developer can get a quick status of the local adaptations (and do the evaluation to see what needs to be upstreamed back)
* To recover the past history, the developer needs to fech the `ref/build/XX` branches. Depending on how "XX" is defined, we can revisit the branch "build/Tumbleweed-$DATE", or "build/Tubleweed-$VERSION-$RELEASE".  In that way the force-push is only to tidy the current representation, without loosing the past history.

Something bad about this is not a model as simple as I would like.  Maybe there are better ideas here about how to fulfill those two opposite goals: (a) clear separation of downstream / upstream commits (without interleaving history), (b) keeping all the history of changes in an accessible way that any developer can revisit and diff.

Can be that (a) is not as important as I think, tho.

**Comment by lnussel** [comment]
I had an idea for a potential solution the other day. I've described  it in #7
